Pattern matcher improvements

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97741
* #97931
* __->__ #97740

This adds support for multi-output patterns and example-based
replacements.

Tests/usage are next in this PR stack.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

jansel committed:
Pattern matcher improvements

This adds support for multi-output patterns and example-based
replacements.

Tests/usage are next in this PR stack.

[ghstack-poisoned]

jansel committed:
Update on "Pattern matcher improvements"


This adds support for multi-output patterns and example-based
replacements.

Tests/usage are next in this PR stack.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "Pattern matcher improvements"


This adds support for multi-output patterns and example-based
replacements.

Tests/usage are next in this PR stack.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "Pattern matcher improvements"


This adds support for multi-output patterns and example-based
replacements.

Tests/usage are next in this PR stack.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "Pattern matcher improvements"


This adds support for multi-output patterns and example-based
replacements.

Tests/usage are next in this PR stack.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "Pattern matcher improvements"


This adds support for multi-output patterns and example-based
replacements.

Tests/usage are next in this PR stack.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -404,7 +580,117 @@ def visit(other_node):
+@torch.no_grad()
+def inference_graph(fn, args):
+    """Build a normalized inference graph, for use with fx_to_pattern"""
+    gm = make_fx(fn, select_decomp_table())(*args)
```
Do we need to maintain any invariants here about functionalization ?

eellison replied:
We'll need to do the RNG functionalization at least I think right ?

jansel replied:
We will need the behavior here to match AotAutograd for the patterns being used.

eellison reviewed (commented):
Cool ! Couple questions

jansel commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -285,107 +387,181 @@ def apply(self, match: Match, graph: torch.fx.Graph, node: torch.fx.Node):
+                elif new is None:
+                    old.replace_all_uses_with(None)
+                else:
+                    new.meta.update(old.meta)
```
I'll add something.

jansel commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -404,7 +580,117 @@ def visit(other_node):
+    )(*args)
+
+    # remove in/out specs
+    gm.graph._codegen = torch.fx.graph.CodeGen()
```
These are the same args that compile_fx() uses, so it should get functionalization.

jansel committed:
Update on "Pattern matcher improvements"


This adds support for multi-output patterns and example-based
replacements.

Tests/usage are next in this PR stack.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented:
@pytorchbot merge

jansel commented:
@pytorchbot merge
